controller/lib/hal/flash.cpp

```diff
@@ -0,0 +1,117 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#if defined(BARE_STM32)
+
+#include "flash.h"
+#include "hal.h"
+#include "hal_stm32.h"
+
+/*
+ * Erase a single 2k page of flash given the starting
+ * address of that page.
+ *
+ * Returns true on success or false on failure
+ */
+bool HalApi::FlashErasePage(uint32_t addr) {
+
+  // Check for a reasonable address
+  if (addr < flash_start_addr)
+    return false;
+
+  if (addr >= flash_start_addr + flash_size)
+    return false;
+
+  // Address needs to be the start of a page
+  if (addr & (flash_page_size - 1))
+    return false;
+
+  // Clear all the status bits
+  FlashReg *reg = FLASH_BASE;
+  reg->status = 0x0000C3FB;
+
+  // Unlock flash by writing two special values to the key register
+  reg->key = 0x45670123;
+  reg->key = 0xCDEF89AB;
+
+  // Find the page number
+  uint8_t n = static_cast<uint8_t>((addr - flash_start_addr) / flash_page_size);
+
+  reg->ctrl.page = n;
+  reg->ctrl.page_erase = 1;
+  reg->ctrl.start = 1;
+
+  // Wait for the busy bit to clear
+  while (reg->status & 0x00010000) {
+  }
+
+  // Lock the flash again
+  reg->ctrl.page = 0;
+  reg->ctrl.page_erase = 0;
+  reg->ctrl.lock = 1;
+  return true;
+}
+
+// Write ct 32-bit integers to flash starting at the given address.
+// ct must be even since flash is programmed 64-bits at a time.
+// The address must be a multiple of 8 for the same reason
+bool HalApi::FlashWrite(uint32_t addr, uint32_t *data, int ct) {
```

Comment:
If I'm understanding the code that uses this, we can program an arbitrary 64-bit address to 0, but if we want to set any 1 bits, it will fail (silently? loudly?) if the block wasn't just erased?

That's probably obvious to anyone who uses flash regularly, but perhaps we could write it down somewhere just in case?  It's critical to the correctness of the `Invalidate()` function.

Reply:
Yup, the way flash works you need to erase it before you can write.  Erasing flash sets all bits to 1.
You can then program bits to 0, but you can't set a 0 bit back to 1 without erasing the whole page.